### PR TITLE
Scintilla 5.1.5 aftermath

### LINKED ISSUE #3095: Crash after update to Scintilla 5.1.5
After the update to Scintilla 5.1.5, I started getting memory corruption crashes when saving some files. (Sorry for not noticing this before, I thought the Scintilla update didn't contain any changes on the Geany side.) I may be wrong but I think the problem is in

https://github.com/geany/geany/pull/3046/commits/d7c985e47412f8a53c5a8202a0f8b669c4fd5e08

and misunderstanding if the following Scintilla API change

>When calling SCI_GETTEXT, SCI_GETSELTEXT, and SCI_GETCURLINE with a NULL buffer argument to discover the length that should be allocated, do not include the terminating NUL in the returned value. The value returned is 1 less than previous versions of Scintilla. Applications should allocate a buffer 1 more than this to accommodate the NUL. The wParam (length) argument to SCI_GETTEXT and SCI_GETCURLINE also omits the NUL. This is more consistent with other APIs.

This part

>When calling SCI_GETTEXT, SCI_GETSELTEXT, and SCI_GETCURLINE with a NULL buffer argument to discover the length that should be allocated, do not include the terminating NUL in the returned value...

> ...The wParam (length) argument to SCI_GETTEXT and SCI_GETCURLINE also omits the NUL.

says that that the returned __length__ doesn't include NUL, not that the returned string misses the NUL character. The rest of the change description says that we should allocate `+1` buffer to accommodate for that.

I believe we should revert the above patch and then go through all the uses of `SCI_GETTEXT`, `SCI_GETSELTEXT`, and `SCI_GETCURLINE` and corresponding `sci_` wrappers and update the code where needed (including plugins). I don't think it's a good idea to update the `sci_` wrappers alone to mirror previous behavior since the whole Scintilla API is part of Geany's API so this is a API change anyway and making `sci_` functions do something else than direct Scintilla calls is confusing.

Rant: I really don't understand why Neil made this change - even though it fixes some inconsistency in Scintilla API, this is a hard to spot backwards incompatible change and such things shouldn't be made in minor releases of a library.

@kugel- @eht16 @elextr What do you think?

### PR DESCRIPTION
**Pull request by kugel-**
This commit fixes a few problems introduced by the last Scintilla update.
That update caused some headache around the incompatible changes to
`SCI_GETTEXT`, `SCI_GETSELTEXT`, and `SCI_GETCURLINE`.

- An explicit NUL termination was added to `sci_get_text()`. This is both
  superflous and wrong (it writes behind the allocated buffer) as SCI_GETTEXT
  already does NUL termination.

- In `sci_get_contents()`, sci_get_string() cannot be used. That would call
  SCI_GETTEXT with length == 0 which is not the desired outcome. Instead,
  basically revert to the old implementation but account for the API change.

- The callers of sci_get_selected_text_length() must be adapted, this
  was missing yet. sci_get_selected_text_length() return value does not
  include the NUL termination anymore.

Resolves #3095

Fixes: d7c985e47 ("Adapt to SCI_GETTEXT changes")

### CODE REVIEW

**Comment by techee** [comment]
Just went through the patch and looks good to me. This should be basically all needed for Geany itself.

**Comment by kugel-** [comment]
Thanks. I'll merge quickly as it resolves pretty severe issue.

**Comment by techee** [comment]
> Thanks. I'll merge quickly as it resolves pretty severe issue.

Pity. What a pleasant change to see that Geany is capable of crashing ;-).
